### ISSUE TITLE
Feature request: depend on `external_program`

### ISSUE THREAD

**Issue by klokik** [mesonbuild/meson#13877]
One can put an `external_program` returned by the `find_program()` as a dependency for the custom target only if the program is an executable target that is a part of the built (so it is likely an `exe` not an `external_program`). Unfortunately, you cannot depend on the program found in the PATH.

Moreover, I didn't see a clean solution for determining where the program is coming from that covers all the cases.
It would be nice to unify the experience and always put the program into the list of dependencies.

Note: it is about the case not covered by auto-added dependencies in the list of the `command` parameters.
Consider the following case:
~~~
protoc = find_program('protoc')
grpc_cpp_plugin = find_program('grpc_cpp_plugin')

proto_generated_sources = custom_target('my-proto-gen-sources',
  input: proto_files,
  output: proto_generated_output_names,
  command: [protoc,
    '--proto_path=@CURRENT_SOURCE_DIR@',
    '--plugin=protoc-gen-grpc=' + grpc_cpp_plugin.full_path(), # this cannot be used to auto add a dependency
    '--grpc_out=@OUTDIR@', '--cpp_out=@OUTDIR@',
    '@INPUT@'],
  depends: [is_grpc_internal ? grpc_cpp_plugin : []], # how to determine `is_grpc_internal` is not clear
)
~~~

**Comment by dcbaker** [comment]
This is a well known issue that we've talked about how to handle it, because we *don't* want to add it to depends, we want to handle it correctly via command.

I had proposed https://github.com/mesonbuild/meson/pull/12287 at one point (now very out of date), but that didn't really move forward.

**Comment by dcbaker** [comment]
Generally, whenever `.full_path()` is getting called, that's a code smell.

**Comment by klokik** [comment]
Will handling it via command lead to removal of the `depends` parameter?


So far, I find this behavior being inconsistent.

ср, 6 лист. 2024 р., 21:51 користувач Dylan Baker ***@***.***>
пише:

**Comment by Ericson2314** [comment]
I am understanding this correctly coming from https://github.com/mesonbuild/meson/issues/14874, the thing we would like is the simple-stupid thing of just filtering out external programs in dependency lists, since they are trivially up to date. (If the external world did change, Ninja isn't expected to know, you need to instead reconfigure.)

**Comment by dcbaker** [comment]
I still like the idea of having a way of doing what #12287 set out to do, because I still see that as inherently useful, but that doesn't preclude having better handling of programs in `depends`

Given that the output of `find_program()` can be three things:
1. A program provided outside of Meson
2. An `executable` target that Meson builds
3. A script that is part of the repo

2 and 3 should have the ability to trigger a rebuild, even if they are not passed as explicit arguments. (and indeed, 3 already can through depend_files for targets)

**Comment by Ericson2314** [comment]
Yes, I think I agree with all that. I wouldn't want a quick fix that screws over #12277 either :).
